File: code/typescript-for-csharp-java/compare/l03_cast.cs
// compare/l03_cast.cs
object parsed = "not a camera";
try { var camera = (CameraState)parsed; Console.WriteLine(camera); }
catch (InvalidCastException e) { Console.WriteLine($"(CameraState)parsed: {e.GetType().Name}"); }
Console.WriteLine($"parsed as CameraState: {(parsed as CameraState) is null}");
Console.WriteLine($"parsed is CameraState: {parsed is CameraState}");

record CameraState(double Px, double Py, double Pz);

File: code/typescript-for-csharp-java/compare/L03Cast.java
// compare/L03Cast.java
record CameraState(double px, double py, double pz) {}

public class L03Cast {
    public static void main(String[] args) {
        Object parsed = "not a camera";
        try {
            var camera = (CameraState) parsed;
            System.out.println(camera);
        } catch (ClassCastException e) {
            System.out.println("(CameraState) parsed: " + e.getClass().getSimpleName());
        }
        if (parsed instanceof CameraState camera) {
            System.out.println(camera.pz());
        } else {
            System.out.println("parsed instanceof CameraState: false");
        }
    }
}
